server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -105,13 +104,13 @@ public String description() {
     }
 
     @Override
-    public void notifyListener(OpenSearchException e, boolean sendShardFailure) {
+    public void notifyListener(ReplicationFailedException e, boolean sendShardFailure) {
         // Cancellations still are passed to our SegmentReplicationListner as failures, if we have failed because of cancellation
         // update the stage.
         final Throwable cancelledException = ExceptionsHelper.unwrap(e, CancellableThreads.ExecutionCancelledException.class);
         if (cancelledException != null) {
             state.setStage(SegmentReplicationState.Stage.CANCELLED);
-            listener.onFailure(state(), (CancellableThreads.ExecutionCancelledException) cancelledException, sendShardFailure);
+            listener.onFailure(state(), new ReplicationFailedException(indexShard, cancelledException), sendShardFailure);
```

Comment:
A nit - I don't think we need to cast or re-wrap the exception here. 

```java
    @Override
    public void notifyListener(ReplicationFailedException e, boolean sendShardFailure) {
        // Cancellations still are passed to our SegmentReplicationListner as failures, if we have failed because of cancellation
        // update the stage.
        if (ExceptionsHelper.unwrap(e, CancellableThreads.ExecutionCancelledException.class) != null) {
            state.setStage(SegmentReplicationState.Stage.CANCELLED);
        }
        listener.onFailure(state(), e, sendShardFailure);
}
```